File: web/src/lib/protocol/crypto.ts
import { CONTENT_HASH_LABEL, KEY_CHECK_LABEL, KEY_LENGTH, NONCE_LENGTH, SEAL_OVERHEAD } from './constants';
import { type Bytes, randomBytes, toHex, utf8 } from './encoding';

export class DecryptError extends Error {
  constructor(message = 'decryption failed') {
    super(message);
    this.name = 'DecryptError';
  }
}

export interface KeySet {
  aes: CryptoKey;
  contentHash: CryptoKey;
  keyCheck: string;
}

function subtle(): SubtleCrypto {
  const s = globalThis.crypto?.subtle;
  if (!s) throw new Error('WebCrypto is not available. Use HTTPS or localhost.');
  return s;
}

export function normalizePassword(password: string): Bytes {
  return utf8(password.normalize('NFC'));
}

export async function deriveMasterKeyBits(password: string, salt: Uint8Array, iterations: number): Promise<Bytes> {
  const base = await subtle().importKey('raw', normalizePassword(password), 'PBKDF2', false, ['deriveBits']);
  const bits = await subtle().deriveBits(
    { name: 'PBKDF2', hash: 'SHA-256', salt: new Uint8Array(salt), iterations },
    base,
    KEY_LENGTH * 8
  );
  return new Uint8Array(bits);
}

export async function hmacSha256(keyBytes: Uint8Array, data: Uint8Array): Promise<Bytes> {
  const k = await subtle().importKey('raw', new Uint8Array(keyBytes), { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']);
  return new Uint8Array(await subtle().sign('HMAC', k, new Uint8Array(data)));
}

export interface SubkeyBytes {
  contentHashKey: Bytes;
  keyCheck: string;
}

export async function deriveSubkeyBytes(master: Uint8Array): Promise<SubkeyBytes> {
  const contentHashKey = await hmacSha256(master, utf8(CONTENT_HASH_LABEL));
  const keyCheck = toHex(await hmacSha256(master, utf8(KEY_CHECK_LABEL)));
  return { contentHashKey, keyCheck };
}

export async function importKeySet(master: Uint8Array, extractable = false): Promise<KeySet> {
  if (master.length !== KEY_LENGTH) throw new Error('key must be 32 bytes');
  const aes = await subtle().importKey('raw', new Uint8Array(master), { name: 'AES-GCM' }, extractable, ['encrypt', 'decrypt']);
  const sub = await deriveSubkeyBytes(master);
  const contentHash = await subtle().importKey('raw', sub.contentHashKey, { name: 'HMAC', hash: 'SHA-256' }, extractable, [
    'sign'
  ]);
  sub.contentHashKey.fill(0);
  return { aes, contentHash, keyCheck: sub.keyCheck };
}

export async function deriveKeySet(password: string, salt: Uint8Array, iterations: number): Promise<KeySet> {
  const master = await deriveMasterKeyBits(password, salt, iterations);
  try {
    return await importKeySet(master, false);
  } finally {
    master.fill(0);
  }
}

export function aadMeta(id: string): Bytes {
  return utf8(`yc1|meta|${id}`);
}

export function aadPayload(id: string): Bytes {
  return utf8(`yc1|payload|${id}`);
}

export function aadThumb(id: string): Bytes {
  return utf8(`yc1|thumb|${id}`);
}

export function aadChunk(id: string, index: number, count: number): Bytes {
  return utf8(`yc1|chunk|${id}|${index}|${count}`);
}

export async function seal(key: CryptoKey, aad: Uint8Array, plaintext: Uint8Array, nonce?: Uint8Array): Promise<Bytes> {
  const iv = nonce ? new Uint8Array(nonce) : randomBytes(NONCE_LENGTH);
  if (iv.length !== NONCE_LENGTH) throw new Error('nonce must be 12 bytes');
  const ct = new Uint8Array(
    await subtle().encrypt(
      {
        name: 'AES-GCM',
        iv,
        additionalData: new Uint8Array(aad),
        tagLength: 128
      },
      key,
      plaintext as Bytes
    )
  );
  const out = new Uint8Array(NONCE_LENGTH + ct.length);
  out.set(iv, 0);
  out.set(ct, NONCE_LENGTH);
  return out;
}

export async function open(key: CryptoKey, aad: Uint8Array, sealed: Uint8Array): Promise<Bytes> {
  if (sealed.length < SEAL_OVERHEAD) throw new DecryptError('sealed box is too short');
  const iv = sealed.slice(0, NONCE_LENGTH);
  const body = sealed.slice(NONCE_LENGTH);
  try {
    const pt = await subtle().decrypt(
      {
        name: 'AES-GCM',
        iv,
        additionalData: new Uint8Array(aad),
        tagLength: 128
      },
      key,
      body
    );
    return new Uint8Array(pt);
  } catch {
    throw new DecryptError();
  }
}

export async function importAesKeyRaw(raw: Uint8Array): Promise<CryptoKey> {
  return subtle().importKey('raw', new Uint8Array(raw), { name: 'AES-GCM' }, false, ['encrypt', 'decrypt']);
}

export async function contentHash(hashKey: CryptoKey, content: Uint8Array): Promise<string> {
  return toHex(new Uint8Array(await subtle().sign('HMAC', hashKey, content as Bytes)));
}

export async function sha256Hex(data: Uint8Array): Promise<string> {
  return toHex(new Uint8Array(await subtle().digest('SHA-256', data as Bytes)));
}

File: protocol/tools/genvectors/crypto.go
package main

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/pbkdf2"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"strconv"
)

const (
	protocolVersion   = 1
	kdfIterations     = 600000
	fastKDFIterations = 1000
	keyLength         = 32
	nonceLength       = 12
	tagLength         = 16
	sealOverhead      = nonceLength + tagLength
	inlineMaxBytes    = 262144
	chunkSizeBytes    = 4194304
	previewMaxPoints  = 500
	archiveMagic      = "YCF1"
)

var (
	contentHashLabel = []byte("yikz-clipboard/v1/content-hash")
	keyCheckLabel    = []byte("yikz-clipboard/v1/key-check")
)

func must(err error) {
	if err != nil {
		panic(err)
	}
}

func deriveKey(password string, salt []byte, iterations int) []byte {
	key, err := pbkdf2.Key(sha256.New, password, salt, iterations, keyLength)
	must(err)
	return key
}

func hmacSHA256(key, message []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write(message)
	return mac.Sum(nil)
}

func contentHashKey(key []byte) []byte {
	return hmacSHA256(key, contentHashLabel)
}

func keyCheck(key []byte) string {
	return hex.EncodeToString(hmacSHA256(key, keyCheckLabel))
}

func contentHash(key, content []byte) string {
	return hex.EncodeToString(hmacSHA256(contentHashKey(key), content))
}

func newGCM(key []byte) cipher.AEAD {
	block, err := aes.NewCipher(key)
	must(err)
	gcm, err := cipher.NewGCM(block)
	must(err)
	return gcm
}

func seal(key, nonce, aad, plaintext []byte) []byte {
	if len(nonce) != nonceLength {
		panic("bad nonce length")
	}
	out := append([]byte{}, nonce...)
	return newGCM(key).Seal(out, nonce, plaintext, aad)
}

func open(key, sealed, aad []byte) ([]byte, error) {
	if len(sealed) < sealOverhead {
		return nil, fmt.Errorf("sealed box too short")
	}
	return newGCM(key).Open(nil, sealed[:nonceLength], sealed[nonceLength:], aad)
}

func aadMeta(id string) []byte {
	return []byte("yc1|meta|" + id)
}

func aadPayload(id string) []byte {
	return []byte("yc1|payload|" + id)
}

func aadThumb(id string) []byte {
	return []byte("yc1|thumb|" + id)
}

func aadChunk(id string, index, count int) []byte {
	return []byte("yc1|chunk|" + id + "|" + strconv.Itoa(index) + "|" + strconv.Itoa(count))
}

func testBytes(label string, n int) []byte {
	var out []byte
	for counter := uint32(0); len(out) < n; counter++ {
		var c [4]byte
		binary.BigEndian.PutUint32(c[:], counter)
		sum := sha256.Sum256(append([]byte("yikz-clipboard test bytes|"+label+"|"), c[:]...))
		out = append(out, sum[:]...)
	}
	return out[:n]
}

func testNonce(label string) []byte {
	return testBytes("nonce|"+label, nonceLength)
}

func uuidV7(unixMillis int64, random []byte) string {
	if len(random) != 10 {
		panic("uuidv7 needs 10 random bytes")
	}
	var b [16]byte
	var ts [8]byte
	binary.BigEndian.PutUint64(ts[:], uint64(unixMillis))
	copy(b[0:6], ts[2:8])
	copy(b[6:16], random)
	b[6] = 0x70 | (b[6] & 0x0f)
	b[8] = 0x80 | (b[8] & 0x3f)
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}

func chunkCount(size int64) int {
	if size <= inlineMaxBytes {
		return 0
	}
	return int((size + chunkSizeBytes - 1) / chunkSizeBytes)
}

func chunkPlainSize(size int64, index, count int) int64 {
	if index < count-1 {
		return chunkSizeBytes
	}
	return size - int64(count-1)*chunkSizeBytes
}

type archiveFile struct {
	Name string
	Data []byte
}

func packFiles(files []archiveFile) []byte {
	var buf bytes.Buffer
	buf.WriteString(archiveMagic)
	var u32 [4]byte
	var u64 [8]byte
	binary.BigEndian.PutUint32(u32[:], uint32(len(files)))
	buf.Write(u32[:])
	for _, f := range files {
		name := []byte(f.Name)
		binary.BigEndian.PutUint32(u32[:], uint32(len(name)))
		buf.Write(u32[:])
		buf.Write(name)
		binary.BigEndian.PutUint64(u64[:], uint64(len(f.Data)))
		buf.Write(u64[:])
		buf.Write(f.Data)
	}
	return buf.Bytes()
}

func archiveSize(files []archiveFile) int64 {
	total := int64(8)
	for _, f := range files {
		total += 4 + int64(len(f.Name)) + 8 + int64(len(f.Data))
	}
	return total
}

func preview(s string) string {
	runes := []rune(s)
	if len(runes) > previewMaxPoints {
		runes = runes[:previewMaxPoints]
	}
	return string(runes)
}

func b64(b []byte) string {
	return base64.StdEncoding.EncodeToString(b)
}

func hexs(b []byte) string {
	return hex.EncodeToString(b)
}

func sha256Hex(b []byte) string {
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:])
}
